templates/rust-api.rs.j2

```diff
@@ -0,0 +1,1290 @@
+//! SWI call, and cannot be safely described to Rust at compile time.
+//!
+//! All SWI calls use the X form of the SWI, and return a Result enum to allow
+//! idiomatic Rust error handling.
+//!
+//! Structs or unions can be created with new() or default().
+//!
+//! Structs with a variable length array as the last member are treated as Rust
+//! dynamically sized types. You cannot create one of these directly, but you
+//! can create a base type with a fixed size that can then be coerced to a
+//! reference to a dynamically sized type. In some cases, an iterator is
+//! provided to safely access the dynamically sized array.
+//!
+//! Unions that are embedded inside another struct can be safely accessed
+//! using provided methods. Reading a union member will first check that the
+//! appropriate struct flags are consistent with that member, and will return
+//! None otherwise. Writing to a union member via a mutable reference from the
+//! access function will set the appropriate flags to indicate that that union
+//! member is the only valid one now.
+//!
+//! # Examples
+//!
+//! ```
+//! # fn main() {
+//! // Create a window with no icons
+//! use oslib::wimp::*;
+//!
+//! let mut block = WimpWindowBase::<0>::new();
+//!
+//! // Set flags where the default value is not sufficient
+//! block.highlight_bg = WIMP_COLOUR_CREAM;
+//! block.title_flags = WIMP_ICON_VCENTRED | WIMP_ICON_RJUSTIFIED;
+//!
+//! let mut title: Vec<u8> = Vec::from(c"WindowTitle".to_bytes_with_nul());
+//! // Access union member, automatically setting the WIMP_ICON_INDIRECTED
+//! // and WIMP_ICON_TEXT title_flags
+//! block.title_data_indirected_text_mut().text = title.as_mut_ptr();
+//! block.title_data_indirected_text_mut().size = title.len() as i32;
+//! block.title_data_indirected_text_mut().validation = std::ptr::null();
+//!
+//! let window_handle = wimp_create_window(&block).unwrap();
+//! # }
+//! ```
+//!
+//! ```
+//! # fn main() {
+//! // Load buffer, e.g. from a file
+//! let mut buffer: Vec<u8> = [0_u8; 120].into();
+//!
+//! use oslib::wimp::*;
+//!
+//! // Obtain a reference of the correct type to the loaded data
+//! let block = WimpWindow::from_bytes(buffer.as_slice());
+//!
+//! // Display all the icons present in the window block
+//! for icon in block {
+//!   println!("Icon: {:#?}", icon);
+//! }
+//!
+//! let window_handle = wimp_create_window(block).unwrap();
+//! # }
+//! ```
+
+{%- for mod in defmod.needs %}
+#[allow(unused)]
+use crate::{{ mod|lower }}::*;
+{%- endfor %}
+
+{# Define information that is not present in the defmod files #}
+
+{# Define what determines when a union field is valid.
+   {'Enclosing struct':
+    {'Struct field that holds the union':
+     {'Union field':
+      ['Struct field that defined validity', 'Mask value for field, or None', 'Match value for field']}}}
+#}
+{% set union_discriminants = {'Alarm_MessageSet': {'time': {'utc': ['set_reason', None, 'ALARM_REASON_SET_UTC'],
```

Comment:
Neat. I wonder if maybe we moved this to a JSON file, so that the data for this could be updated without having to worry about Jinja2. Not for now as I think it's worth getting something in and then we can worry about how to structure better, if that's even needed.